### ISSUE TITLE
Bands workchain can't be called without relax when using a builder

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida-quantumespresso#401]
**Steps to reproduce:**
Create a ``PwBandsWorkChain`` using the process builder, without specifying any ``relax`` inputs.

**Expected behavior:**
The workchain skips the relaxation step.

**Actual behavior:**
The workchain fails to validate its inputs, claiming a missing ``code`` input for the ``relax`` part.

**Notes:**
The issue seems to come from the fact that the builder contains
```
'relax': {'metadata': {}, 'base': {'metadata': {}, 'pw': {'metadata': {'options': {}}, 'pseudos': {}}}}
```
Deleting the ``relax`` attribute entirely isn't possible, and replacing it with ``None`` or ``{}`` doesn't fix the issue.

Creating the same input as a dictionary works as expected.

I'm not sure if this is a problem with ``aiida-core``, the way the spec is built in ``aiida-quantumespresso``, or kind of inevitable. @sphuber is the expert on the topic in any case :smile:

**Comment by sphuber** [comment]
I was wondering when anyone was going to run into this one! :smile: 

The problem stems from the way the process spec validation is performed. When passing a mapping of inputs to construct a process, the inputs are first pre-processes, by mapping them onto the spec to populate any defaults that are missing. Since this has to be done recursive, the nested mappings are created on the way. However, the automatic population of defaults was causing problems with the validation. For example, for a spec with an optional namespace, that within it maybe contained a port with a default, could no longer be omitted. Even when not specifying the namespace, the default population step would create it anyway and populate the port with the default. The presence of the namespace would trigger its validation which would then subsequently fail for any other ports that don't have a default.

To tackle this problem, I introduced the [concept of `lazy` namespaces in `plumpy`](https://github.com/aiidateam/plumpy/commit/99c100a78574bf644b4d97f74f2eb9b698ee14e4). This will allow to mark a port namespace with `populate_defaults=False` such that optional namespaces whose namespace is not explicitly defined, will not be pre-processed and therefore subsequent validation will be skipped. Even though that code is already released and required by `aiida-core` I cannot find the argument being used anywhere in the code. It has been a few months since I looked into this and am not sure where I left things.

The process builder is another separate problem, because that will also by default create all the available nested namespaces. So even if you don't specify anything in the `relax` namespace, when submitting it, like you noticed, it does indeed have `relax: {}`, which again triggers pre-processing and then a failing validation. I have a fix in my git stash to prune "empty" nested namespaces, that I wrote at the time to fix this:
```
def prune(value):
    import collections

    if isinstance(value, collections.Mapping):
        result = {}
        for key, sub_value in value.items():
            pruned = prune(sub_value)
            if pruned is None or pruned:
                result[key] = pruned
        return result

    if isinstance(value, collections.Sequence):
        result = []
        for sub_value in value:
            pruned = prune(sub_value)
            if pruned is None or pruned:
                result.append(pruned)
        return result

    return value
```
which I can then add in the `aiida/engine/utils/instantiate_process` for the `ProcessBuilder` case. This works, but I am not convinced of the solution so I never put this in `aiida-core`.

**Comment by greschd** [comment]
Ok, I think this is a discussion for ``aiida-core``. The port namespaces here _do_ use ``populate_defaults=True``, which seems like the correct way of doing it -- now the question is what happens with that on the side of ``aiida-core``.

> To tackle this problem, I introduced the [concept of `lazy` namespaces in `plumpy`](https://github.com/aiidateam/plumpy/commit/99c100a78574bf644b4d97f74f2eb9b698ee14e4). This will allow to mark a port namespace with `populate_defaults=False` such that optional namespaces whose namespace is not explicitly defined, will not be pre-processed and therefore subsequent validation will be skipped. Even though that code is already released and required by `aiida-core` I cannot find the argument being used anywhere in the code. It has been a few months since I looked into this and am not sure where I left things.

I think we don't run into this here because the ``relax`` namespace doesn't have defaults -- otherwise the dict approach also wouldn't work, right?

What's the reason for also pruning empty sequences instead of only mappings? That would also catch an empty AiiDA ``List``. I'm not sure if that (or an empty dict) might ever have meaning. I'm not at all familiar with the builder implementation, but maybe we could construct these "lazy" namespaces only when accessed (``__getattr__`` magic)?

**Comment by greschd** [comment]
Ah that might mess with tab-completion.. another option would be to somehow flag the lazy namespace (wrap it in a different object?) and then replace it with a normal one as soon as it's modified.

**Comment by greschd** [comment]
@sphuber can you transfer this issue to ``aiida-core``?

**Comment by sphuber** [comment]
> The port namespaces here _do_ use `populate_defaults=True`, which seems like the correct way of doing it

Yes, but in the `PwBandsWorkChain` I pass `populate_defaults=False` when exposing the inputs of `PwRelaxWorkChain`. With the combination that the `relax` namespace is `required=False`, it allows the caller not to pass the namespace at all. This allows the heuristic of `PwBandsWorkChain` to make the relax part optional by checking if the namespace was passed in the inputs. This works if one builds the inputs dictionary itself. It just doesn't work when using the `ProcessBuilder` because it will always have an empty mapping, even if nothing was explicitly set.

So far we are on the same page?

What should the solution be then in your opinion? I am not sure if it is possible to change the builders implementation to not create placeholder namespaces if they are not set explicitly. Would the prune functionality be a valid solution too? I feel it is a bit dirty, which is probably why I never opened a PR for it on `aiida-core`. What do you think?

**Comment by greschd** [comment]
> 
> Yes, but in the `PwBandsWorkChain` I pass `populate_defaults=False` when exposing the inputs of `PwRelaxWorkChain`. With the combination that the `relax` namespace is `required=False`, it allows the caller not to pass the namespace at all. This allows the heuristic of `PwBandsWorkChain` to make the relax part optional by checking if the namespace was passed in the inputs. This works if one builds the inputs dictionary itself. It just doesn't work when using the `ProcessBuilder` because it will always have an empty mapping, even if nothing was explicitly set.
> 
> So far we are on the same page?

Yeah, I think we agree here -- if I understand you right this just needs double-checking if the ``lazy`` concept of plumpy is actually used in ``aiida-core``.

> What should the solution be then in your opinion? I am not sure if it is possible to change the builders implementation to not create placeholder namespaces if they are not set explicitly. Would the prune functionality be a valid solution too? I feel it is a bit dirty, which is probably why I never opened a PR for it on `aiida-core`. What do you think?

I think pruning could work.. maybe it needs a bit of refinement to make sure it doesn't accidentally transform legitimate values. Couldn't we just have it trigger for ``ProcessBuilderNamespace`` only?

**Comment by greschd** [comment]
Agree that placeholder namespaces are difficult btw. -- without wrapping everything in descriptors all the way down (which would be a mess) it's kind of impossible to detect if anything in the namespace is changed.

**Comment by sphuber** [comment]
> if I understand you right this just needs double-checking if the `lazy` concept of plumpy is actually used in `aiida-core`.

No, this is exactly the point (and why I was a bit confused when I didn't find it used in `aiida-core`) but this feature should not be used by default. Essentially this is to be used for optional port namespaces, i.e. when creating one with both `required=False` and `populate_defaults=False` as I do for the `relax` namespace in `PwBandsWorkChain`. That is why it is not used in `aiida-core`.

However, the problem in combination with the `ProcessBuilder` _is_ I think something that should probably be fixed in `aiida-core`. So I think I will move this issue to `aiida-core` and suggest the solution to use pruning of builder input mapping before constructing the process. But like you said, my proposed snippet will need some scrutiny and refinement to ensure that valid "emptyish" values are not pruned.
